lib/query-interface.js

```diff
@@ -495,18 +495,20 @@ QueryInterface.prototype.insert = function(instance, tableName, values, options)
   });
 };
 
-QueryInterface.prototype.upsert = function(tableName, values, updateValues, model, options) {
+QueryInterface.prototype.upsert = function(tableName, valuesByField, updateValues, model, options) {
   var wheres = []
     , where
     , indexFields
     , indexes = []
-    , attributes = Object.keys(values);
+    , attributes = Object.keys(valuesByField);
 
   where = {};
   for (var i = 0; i < model.primaryKeyAttributes.length; i++) {
     var key = model.primaryKeyAttributes[i];
-    if(key in values){
-      where[key] = values[key];
+    var attribute = model.rawAttributes[key];
+    var fieldName = attribute.field || attribute.fieldName;
+    if(fieldName in valuesByField){
+      where[fieldName] = valuesByField[fieldName];
     }
   }
```

Comment:
Consider retaining the original code and simply mapping where through https://github.com/sequelize/sequelize/blob/master/lib/utils.js#L121 before sending it to upsertQuery